app/javascript/src/NewService/components/NewServiceForm.jsx

```diff
@@ -0,0 +1,120 @@
+
+import React, {useState, useEffect} from 'react'
+import {fetch as fetchPolyfill} from 'whatwg-fetch'
+import 'core-js/es6/promise'
+import {ServiceSourceForm, FormTemplate} from 'NewService'
+import type {SourceFormProps, FormTemplateProps} from 'NewService/types'
+
+import {createReactWrapper} from 'utilities/createReactWrapper'
+
+const BASE_PATH = '/p/admin/service_discovery'
+const PROJECTS_PATH = `${BASE_PATH}/projects.json`
+
+type Props = {
+  isServiceDiscoveryAccesible: boolean,
+  isServiceDiscoveryUsable: boolean,
+  serviceDiscoveryAuthenticateUrl: string,
+  providerAdminServiceDiscoveryServicesPath: string,
+  adminServicesPath: string
+}
+
+type Response = {
+  ok: boolean,
+  statusText: string,
+  json: () => Promise<void>
+}
+const fetchData = async (url: string, setError) => {
+  setError({hasError: false, errorMsg: ''})
+  const namespace = url === PROJECTS_PATH
+    ? 'projects'
+    : 'services'
+
+  return fetchPolyfill(url)
+    .then((response: Response) => {
+      if (!response.ok) {
+        setError({hasError: true, errorMsg: response.statusText})
+      }
+      return response.json()
+    })
+    .then(data => data[namespace].reduce((acc, current) => acc.concat(current.metadata.name), []))
+    .catch(error => {
+      console.error(error)
+    })
+}
+
+const callEfect = (servicesRequested, selectedNamespace, startServicesFetch) => {
+  if (servicesRequested && selectedNamespace) {
+    const SERVICES_PATH = `${BASE_PATH}/namespaces/${selectedNamespace}/services.json`
+    startServicesFetch(SERVICES_PATH)
+  }
+}
+
+const renderForm = (sourceFormProps, serviceDiscoveryProps, serviceNewProps) => <React.Fragment>
+  <h1>New API</h1>
+  <ServiceSourceForm {...sourceFormProps}/>
+  <FormTemplate {...serviceDiscoveryProps}/>
+  <FormTemplate {...serviceNewProps}/>
+</React.Fragment>
+
+const onNamespacesFetch = async (setNamespaces, setServicesRequested, settingSelectedNamespace, setError): Promise<void> => {
+  setNamespaces([])
+  const namespacesData = await fetchData(PROJECTS_PATH, setError)
+  if (namespacesData) {
+    setNamespaces(namespacesData)
+    setServicesRequested(true)
+    settingSelectedNamespace(namespacesData[0])
+  }
+}
+
+const onServicesFetch = async (servicesUrl, setServices, setServicesRequested, setError): Promise<void> => {
+  setServices([])
+  const servicesData = await fetchData(servicesUrl, setError)
+  if (servicesData) {
+    setServices(servicesData)
+    setServicesRequested(false)
+  }
+}
+
+const getChangeNamespaces = (event, setSelectedNamespace, setServicesRequested) => {
+  setSelectedNamespace(event.currentTarget.value)
+  setServicesRequested(true)
+}
+
+const getHandleFormsVisibility = (event, setServDescVisible, setServNewVisible) => {
+  if (event.currentTarget.value === 'discover') {
+    setServDescVisible(true)
+    setServNewVisible(false)
+  } else {
+    setServDescVisible(false)
+    setServNewVisible(true)
+  }
+}
+
+const NewServiceForm = (props: Props) => {
```

Comment:
If the methods were defined inside the body of the component it'd be a lot bigger and complex.
I had to pass the state to the methods (isServDescVisible, namespaces, etc.), because using hooks they are not available outside the function body.

Reply:
A way to fix this is what @josemigallas suggested before: `useContext`.
But I'd prefer not, because enzyme still doesn't have full support for hooks, including useContext.
It's true it could help with the props problem, but testing will be more complicated.
As enzyme support for hooks is an [open discussion](https://github.com/airbnb/enzyme/pull/2041), it may change a lot. So I think it's better to wait until we have something more stable.
In the meantime I'd stay with this approach, even if it's more verbose
WDYT @didierofrivia

Reply:
1. You could separate all the fetch in an API-like module, passing the path and scope and you get rid of the methods defined there and the conditional for namespaces. You might also get rid of those `on...` functions and dry them up.
2. Somehow that form manage waaaaaaaay too much, are all those properties really needed to keep a state?

Reply:
AFAIK Enzyme not "supporting" hooks is not an issue, as long as you don't want to change the state. So far we've been testing hook just find simply using props and events and leaving state alone.
In any case, even if the components can't be fully tested now, let's keep in mind that the whole form is being tested in Cucumber so we got our backs covered.